src/Function/Structures.agda

```diff
@@ -0,0 +1,110 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Structures for types of functions
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Relation.Binary
+
+module Function.Structures
+  {a b ℓ₁ ℓ₂} {A : Set a} {B : Set b} (_≈₁_ : Rel A ℓ₁) (_≈₂_ : Rel B ℓ₂)
+  where
+
+open import Data.Product using (∃; _×_; _,_)
+open import Function.Core
+open import Function.Definitions
+open import Level using (_⊔_)
+import Relation.Binary.Reasoning.Setoid as SetoidReasoning
+
+------------------------------------------------------------------------
+-- Definitions
+
+record IsEqualityPreserving (f : A → B) : Set (a ⊔ b ⊔ ℓ₁ ⊔ ℓ₂) where
```

Comment:
Here is where I think things "go wrong". I was expecting `IsEqualityPreserving` to, in a sense, be `IsEqualityPreservingFor`, with two **equivalence relations** (and a function) as parameters. So what seems 'wrong' is that the parameters to `Function.Structures` are just the relations instead of equivalence relations. That seems to be the source of the duplication.

Reply:
> So what seems 'wrong' is that the parameters to Function.Structures are just the relations instead of equivalence relations. That seems to be the source of the duplication.

I agree that this is the source of the duplication, but having the option of only passing the relations and not the equivalence relations  is the main point having the `IsX` variants. 

The problem is that the relation and the carrier may be provided by another package from another hierarchy. For example in #795, the carrier and the equality relation is provided by a `RawMagma`, and therefore under the current scheme it's impossible to neatly use the `_LeftInverseFor_` package because you cannot specify that the carrier and the equality relation in the left inverse match up with the magma. A similar problem would happen if we parameterised the `IsX` modules by a setoid. In contrast the new `IsX` as they stand currently allow you to pass the relation in directly so it all matches up.

Reply:
What about naming this `IsFunction` or `Functional`?

Reply:
I would expect `Functional` to be a property of a relation `R : REL A B r`
stating that for any `a : A`, there is at most one `b : B` such that `R a b`.
(Together with `Total` which states there is at least one such `b`, we get
that `R` corresponds to the graph of a function).

What about `SetoidMorphism`?

Reply:
`SetoidMorphism` seems quite reasonable.

Reply:
I like `IsSetoidMorphism` as well, but isn't it going to be taken when we finally get around to creating a  `Relation.Binary.Morphism` file to match the `Algebra.Morphism` file?

Reply:
`IsSetoidFunction` or `IsSetoidalFunction`?

Reply:
Wouldn't the thing in `Relation.Binary.Morphism` not be identical to this current thing `IsSetoidMorphism` ? If not, what would be the difference?

Reply:
Based on `Algebra.Morphism` I'm guessing it would look like:
```agda
module _ {c₁ ℓ₁ c₂ ℓ₂} (From : Setoid c₁ ℓ₁)  (To   : Setoid c₂ ℓ₂) where

  private
    module F = Setoid From
    module T = Setoid To

  record IsSetoidMorphism (⟦_⟧ : Morphism) :
         Set (c₁ ⊔ ℓ₁ ⊔ c₂ ⊔ ℓ₂) where
    field
      ⟦⟧-cong : ⟦_⟧ Preserves F._≈_ ⟶ T._≈_

  syntax IsSetoidMorphism From To F = F Is From -Setoid⟶ To
```
i.e. it would take setoids as parameters rather than relations.

Reply:
I've got it! @JacquesCarette and @gallais how about `IsCongruent`?

Reply:
Indeed, that does seem to fit rather well.

Reply:
I still consider it good standard practice to attach `syntax` only to single-purpose aliases:
````agda

  record IsSetoidMorphism (⟦_⟧ : Morphism) :
         Set (c₁ ⊔ ℓ₁ ⊔ c₂ ⊔ ℓ₂) where
    field
      ⟦⟧-cong : ⟦_⟧ Preserves F._≈_ ⟶ T._≈_

  IsSetoidMorphismForSyntax : (F : Morphism) → IsSetoidMorphism F

  syntax IsSetoidMorphismForSyntax From To F = F Is From -Setoid⟶ To
````
This makes it possible to “hide the `syntax` import”. (See for example `Σ-syntax` in `Data.Product`.)

Reply:
@WolframKahl nice idea. Hadn't thought about that before, will change it.